File: source/SharpFlame/clsComponent.cs
using SharpFlame.Collections;

namespace SharpFlame
{
    public abstract class clsComponent
    {
        public bool IsUnknown = false;

        public string Name = "Unknown";
        public string Code;

        public enum enumComponentType
        {
            Unspecified,
            Body,
            Propulsion,
            Turret
        }

        public enumComponentType ComponentType = enumComponentType.Unspecified;

        public bool Designable;
    }

    public class clsBody : clsComponent
    {
        public ConnectedListLink<clsBody, clsObjectData> ObjectDataLink;

        public clsUnitType.clsAttachment Attachment = new clsUnitType.clsAttachment();
        public int Hitpoints;

        public clsBody()
        {
            ObjectDataLink = new ConnectedListLink<clsBody, clsObjectData>(this);


            ComponentType = enumComponentType.Body;
        }
    }

    public class clsPropulsion : clsComponent
    {
        public ConnectedListLink<clsPropulsion, clsObjectData> ObjectDataLink;

        public struct sBody
        {
            public clsUnitType.clsAttachment LeftAttachment;
            public clsUnitType.clsAttachment RightAttachment;
        }

        public sBody[] Bodies = new sBody[0];
        public int HitPoints;

        public clsPropulsion(int BodyCount)
        {
            ObjectDataLink = new ConnectedListLink<clsPropulsion, clsObjectData>(this);


            ComponentType = enumComponentType.Propulsion;

            int A = 0;

            Bodies = new sBody[BodyCount];
            for ( A = 0; A <= BodyCount - 1; A++ )
            {
                Bodies[A].LeftAttachment = new clsUnitType.clsAttachment();
                Bodies[A].RightAttachment = new clsUnitType.clsAttachment();
            }
        }
    }

    public class clsTurret : clsComponent
    {
        public clsTurret()
        {
            TurretObjectDataLink = new ConnectedListLink<clsTurret, clsObjectData>(this);
        }

        public ConnectedListLink<clsTurret, clsObjectData> TurretObjectDataLink;

        public clsUnitType.clsAttachment Attachment = new clsUnitType.clsAttachment();
        public int HitPoints;

        public enum enumTurretType
        {
            Unknown,
            Weapon,
            Construct,
            Repair,
            Sensor,
            Brain,
            ECM
        }

        public enumTurretType TurretType = enumTurretType.Unknown;

        public bool GetTurretTypeName(ref string Result)
        {
            switch ( TurretType )
            {
                case enumTurretType.Weapon:
                    Result = "Weapon";
                    return true;
                case enumTurretType.Construct:
                    Result = "Construct";
                    return true;
                case enumTurretType.Repair:
                    Result = "Repair";
                    return true;
                case enumTurretType.Sensor:
                    Result = "Sensor";
                    return true;
                case enumTurretType.Brain:
                    Result = "Brain";
                    return true;
                case enumTurretType.ECM:
                    Result = "ECM";
                    return true;
                default:
                    Result = null;
                    return false;
            }
        }
    }

    public class clsWeapon : clsTurret
    {
        public ConnectedListLink<clsWeapon, clsObjectData> ObjectDataLink;

        public clsWeapon()
        {
            ObjectDataLink = new ConnectedListLink<clsWeapon, clsObjectData>(this);


            TurretType = enumTurretType.Weapon;
        }
    }

    public class clsConstruct : clsTurret
    {
        public ConnectedListLink<clsConstruct, clsObjectData> ObjectDataLink;

        public clsConstruct()
        {
            ObjectDataLink = new ConnectedListLink<clsConstruct, clsObjectData>(this);


            TurretType = enumTurretType.Construct;
        }
    }

    public class clsRepair : clsTurret
    {
        public ConnectedListLink<clsRepair, clsObjectData> ObjectDataLink;

        public clsRepair()
        {
            ObjectDataLink = new ConnectedListLink<clsRepair, clsObjectData>(this);


            TurretType = enumTurretType.Repair;
        }
    }

    public class clsSensor : clsTurret
    {
        public ConnectedListLink<clsSensor, clsObjectData> ObjectDataLink;

        public enum enumLocation
        {
            Unspecified,
            Turret,
            Invisible
        }

        public enumLocation Location = enumLocation.Unspecified;

        public clsSensor()
        {
            ObjectDataLink = new ConnectedListLink<clsSensor, clsObjectData>(this);


            TurretType = enumTurretType.Sensor;
        }
    }

    public class clsBrain : clsTurret
    {
        public ConnectedListLink<clsBrain, clsObjectData> ObjectDataLink;

        public clsWeapon Weapon;

        public clsBrain()
        {
            ObjectDataLink = new ConnectedListLink<clsBrain, clsObjectData>(this);


            TurretType = enumTurretType.Brain;
        }
    }

    public class clsECM : clsTurret
    {
        public ConnectedListLink<clsECM, clsObjectData> ObjectDataLink;

        public clsECM()
        {
            ObjectDataLink = new ConnectedListLink<clsECM, clsObjectData>(this);


            TurretType = enumTurretType.ECM;
        }
    }
}

File: reference/FlaMEVB/clsComponent.vb

Public MustInherit Class clsComponent

    Public IsUnknown As Boolean = False

    Public Name As String = "Unknown"
    Public Code As String
    Public Enum enumComponentType As Byte
        Unspecified
        Body
        Propulsion
        Turret
    End Enum
    Public ComponentType As enumComponentType = enumComponentType.Unspecified

    Public Designable As Boolean
End Class

Public Class clsBody
    Inherits clsComponent

    Public ObjectDataLink As New ConnectedListLink(Of clsBody, clsObjectData)(Me)

    Public Attachment As New clsUnitType.clsAttachment
    Public Hitpoints As Integer

    Public Sub New()

        ComponentType = enumComponentType.Body
    End Sub
End Class

Public Class clsPropulsion
    Inherits clsComponent

    Public ObjectDataLink As New ConnectedListLink(Of clsPropulsion, clsObjectData)(Me)

    Public Structure sBody
        Public LeftAttachment As clsUnitType.clsAttachment
        Public RightAttachment As clsUnitType.clsAttachment
    End Structure
    Public Bodies(-1) As sBody
    Public HitPoints As Integer

    Public Sub New(BodyCount As Integer)

        ComponentType = enumComponentType.Propulsion

        Dim A As Integer

        ReDim Bodies(BodyCount - 1)
        For A = 0 To BodyCount - 1
            Bodies(A).LeftAttachment = New clsUnitType.clsAttachment
            Bodies(A).RightAttachment = New clsUnitType.clsAttachment
        Next
    End Sub
End Class

Public Class clsTurret
    Inherits clsComponent

    Public TurretObjectDataLink As New ConnectedListLink(Of clsTurret, clsObjectData)(Me)

    Public Attachment As New clsUnitType.clsAttachment
    Public HitPoints As Integer
    Public Enum enumTurretType As Byte
        Unknown
        Weapon
        Construct
        Repair
        Sensor
        Brain
        ECM
    End Enum
    Public TurretType As enumTurretType = enumTurretType.Unknown

    Public Function GetTurretTypeName(ByRef Result As String) As Boolean

        Select Case TurretType
            Case enumTurretType.Weapon
                Result = "Weapon"
                Return True
            Case enumTurretType.Construct
                Result = "Construct"
                Return True
            Case enumTurretType.Repair
                Result = "Repair"
                Return True
            Case enumTurretType.Sensor
                Result = "Sensor"
                Return True
            Case enumTurretType.Brain
                Result = "Brain"
                Return True
            Case enumTurretType.ECM
                Result = "ECM"
                Return True
            Case Else
            	Result = Nothing
                Return False
        End Select
    End Function
End Class

Public Class clsWeapon
    Inherits clsTurret

    Public ObjectDataLink As New ConnectedListLink(Of clsWeapon, clsObjectData)(Me)

    Public Sub New()

        TurretType = enumTurretType.Weapon
    End Sub
End Class

Public Class clsConstruct
    Inherits clsTurret

    Public ObjectDataLink As New ConnectedListLink(Of clsConstruct, clsObjectData)(Me)

    Public Sub New()

        TurretType = enumTurretType.Construct
    End Sub
End Class

Public Class clsRepair
    Inherits clsTurret

    Public ObjectDataLink As New ConnectedListLink(Of clsRepair, clsObjectData)(Me)

    Public Sub New()

        TurretType = enumTurretType.Repair
    End Sub
End Class

Public Class clsSensor
    Inherits clsTurret

    Public ObjectDataLink As New ConnectedListLink(Of clsSensor, clsObjectData)(Me)

    Public Enum enumLocation
        Unspecified
        Turret
        Invisible
    End Enum
    Public Location As enumLocation = enumLocation.Unspecified

    Public Sub New()

        TurretType = enumTurretType.Sensor
    End Sub
End Class

Public Class clsBrain
    Inherits clsTurret

    Public ObjectDataLink As New ConnectedListLink(Of clsBrain, clsObjectData)(Me)

    Public Weapon As clsWeapon

    Public Sub New()

        TurretType = enumTurretType.Brain
    End Sub
End Class

Public Class clsECM
    Inherits clsTurret

    Public ObjectDataLink As New ConnectedListLink(Of clsECM, clsObjectData)(Me)

    Public Sub New()

        TurretType = enumTurretType.ECM
    End Sub
End Class
